tools/twix/src/panels/look_at.rs

```diff
@@ -0,0 +1,220 @@
+            Err(error) => {
+                error!("Failed to subscribe: {error:#?}");
+                None
+            }
+        };
+
+        Self {
+            nao,
+            camera_position: Some(CameraPosition::Top),
+            look_at_target: DEFAULT_TARGET,
+            look_at_mode: LookAtType::PenaltyBoxFromCenter,
+            is_enabled: false,
+            field_dimensions,
+            field_dimensions_update_notify_receiver,
+            motion_command,
+        }
+    }
+}
+
+impl Widget for &mut LookAtPanel {
+    fn ui(self, ui: &mut Ui) -> Response {
+        ui.vertical(|ui| {
+            if ui
+                .checkbox(&mut self.is_enabled, "Enable Motion Override")
+                .changed()
+            {
+                if self.is_enabled {
+                    send_standing_look_at(
+                        self.nao.as_ref(),
+                        self.look_at_target,
+                        self.camera_position,
+                    );
+                } else {
+                    self.nao
+                        .update_parameter_value(INJECTED_MOTION_COMMAND, Value::Null);
+                }
+            }
+
+            ui.horizontal(|ui| {
+                ui.vertical(|ui| {
+                    ui.label("Select look-at mode");
+                    ui.radio_value(
+                        &mut self.look_at_mode,
+                        LookAtType::PenaltyBoxFromCenter,
+                        "Look at penalty box from center circle",
+                    );
+                    ui.radio_value(
+                        &mut self.look_at_mode,
+                        LookAtType::Manual,
+                        "Manual target (Robot Coordinates)",
+                    );
+                });
+                ui.vertical(|ui| {
+                    ui.label("Camera to look at with:");
+                    ui.radio_value(
+                        &mut self.camera_position,
+                        Some(CameraPosition::Top),
+                        "Top Camera",
+                    );
+                    ui.radio_value(
+                        &mut self.camera_position,
+                        Some(CameraPosition::Bottom),
+                        "Bottom Camera",
+                    );
+                    ui.radio_value(&mut self.camera_position, None, "Automatic");
+                });
+            });
+
+            let current_field_dimensions = self.field_dimensions.as_ref().and_then(|buffer| {
+                buffer.get_latest().ok().and_then(|latest| {
+                    if self
+                        .field_dimensions_update_notify_receiver
+                        .try_recv()
+                        .is_ok()
+                    {
+                        serde_json::from_value::<FieldDimensions>(latest).ok()
+                    } else {
+                        None
+                    }
+                })
+            });
```

Comment:
Correct, this is partly my inexperience with cleanly unwrapping things and getting them to work, also that the extra option of wrapping the value buffer complicated this even more.

I can reduce the complexity here by making `self.field_dimensions` a `ValueBuffer` instead of an option.

The question you asked in the previous comment about `if let Some(dimensions) = current_field_dimensions {` uses the outcome of this line.

Reply:
This is cleaned up, I might have some misunderstandings on using ` .try_recv()`, if so please let me know.

Reply:
I think listening to updates on the field dimensions is overkill.
You can simply subscribe the parameter as a value buffer and always get the latest value that way.
The update notification is only important if you want to do control flow when an update happens.
But since you're not even sending a new motion command (which is good I think) but only update your locally saved field dimensions, you're essentially reimplementing the value buffer you already have.

For inspiration, look at what the `subscribe` function you are using from the parameter panel actually does.

Reply:
Aha looking at `subscribe` told me the issue. To make this clear;
1. no need to explicitly call `try_recv` -> it is already getting the latest stuff to the buffer.
2. no real* need to store the field_dimension value in the class (IMO as long as I can get the latest value by calling `get_latest()`)

If I understood your explanation, shouldn't the following be sufficient?
```rust
let current_field_dimensions = self.field_dimensions_buffer.get_latest().map_or(None, |latest| {
  serde_json::from_value::<FieldDimensions>(latest).ok();
});
```

Reply:
IIRC there is also a `parse_latest` method which does all of that for you.

Reply:
Didn't know about it, that was quite useful.